### ISSUE TITLE
Feature Request - Remote Control/Usage

### ISSUE THREAD

**Issue by JojoDevel** [pymmcore-plus/pymmcore-remote#3]
## Why?

Our microscopes (mostly NikonTi Eclipse) are controlled by computer just next to them running windows not always on the most up-to-date version and sometimes even air-gapped form the internet. Therefore, we usually try to touch these machines as little as possible, install as little software as possible and let them only interface with the microscope.

## What?

We would like to run the napari-micromanager on any other computer in our lab or institute and attach remotely to the running microscope control computer. This would allow us to much easier maintain/develop software, control imaging loops easier, and maybe even work remotely from our labs and dial into running image acquisitions from all over the world...

## How?

Not sure yet. I imagine a solution with as little programming overhead as possible but keeping the same functionality as right now. So the developer should not feel that everything goes through a remote connection (Though speed will possibly slower).

There is for example, [pyRO](https://pyro5.readthedocs.io/en/latest/) that allows to execute functions remotely. Maybe we could turn every device in the pymmcore python api into such a remote object. And then the python code just uses calls that are executed remotely. On the microscope computer then a pyro server would run initiated from a micromanager configuration. I never used the framework though.

So, there might be also be better opportunities to implement that. Any ideas?

**Hint:** We also discussed that on the EPFL Symposium for Smart Microscopy 2024

**Comment by tlambert03** [comment]
[I moved this issue again here to pymmcore-remote.  I'm sure there will be multiple issues that arise as we test it, and they can hopefully be largely contained in this repo.  Though some changes may need to be made in other repos]

**Comment by JojoDevel** [comment]
> Yes actually i found a good solution with msgpack-numpy and will use it 👍
> 
> I tried briefly to use it with napari micromanager and there are still some remaining issue with serialization of certain custom objects, so I'll keep working on it and let you know when I get a client side gui working with a remote core

This would be pretty amazing for us and that's exactly what we were looking for :star_struck: Then we no longer have to do requests wrapped into a REST API and can use the full ecosystem. But also contribute our devs directly to the napari micromanager gui :)

**Comment by tlambert03** [comment]
Yeah, I agree, it would be a big win all  around for this to be a relative invisible change in the front end.

The biggest issue I'm facing for the gui right now is #4 
Client side signal callbacks that invoke server side core methods aren't working, which is going to be necessary for the gui stuff to work.  I think it's a threading issue, but still need to figure it out

**Comment by tlambert03** [comment]
hey @JojoDevel, here's a promising development.  proof of principle here using @fdrgsp's https://github.com/fdrgsp/micromanager-gui (but could be napari as well) showing 4 different "client-side" applications open.  anything we do in one window is communicated to all the other windows via the server.  So, things are going in a good direction!

https://github.com/user-attachments/assets/3c119901-41b1-4be8-ab71-9cdca6768c3d

**Comment by JojoDevel** [comment]
This looks fantastic :star_struck: :rocket:
